src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs

```diff
@@ -122,30 +133,43 @@ private Block ProduceEmptyBlock(string payloadId, BlockHeader parentHeader, Payl
                         return currentContext;
                     }
 
-                    IBlockImprovementContext newContext = CreateBlockImprovementContext(id, parentHeader, payloadAttributes, currentBestBlock, startDateTime);
+                    IBlockImprovementContext newContext = CreateBlockImprovementContext(id,
+                        parentHeader,
+                        payloadAttributes,
+                        currentBestBlock,
+                        offsetFromStartTime);
+
+                    //TODO: Dispose here may be not thread-safe, possibly investigate atomic swap.
                     currentContext.Dispose();
                     return newContext;
                 });
 
 
-        private IBlockImprovementContext CreateBlockImprovementContext(string payloadId, BlockHeader parentHeader, PayloadAttributes payloadAttributes, Block currentBestBlock, DateTimeOffset startDateTime)
+        private IBlockImprovementContext CreateBlockImprovementContext(
+            string payloadId,
+            BlockHeader parentHeader,
+            PayloadAttributes payloadAttributes,
+            Block currentBestBlock,
+            TimeSpan? offsetFromStartTime)
```

Comment:
It looks to me as if it doesn't do what it did and what it is supposed to do.

So the previous logic was:
1. Take the StartDateTime
2. Run improvement
3. Check if we still have time to run next improvement
4. If yes run next improvement, keep StartDateTime of original(!) improvement
5. Loop to 3.

Now offsetFromStartTime looks not used. We are measuring time only for current Improvement, not all of them.